### PR TITLE
Made RsWithCookie able to combine the given value with previous ones. #293

### PR DESCRIPTION
**Pull request by Bertram25**
Also adapted the corresponding test.

### CODE REVIEW

**Comment by davvd** [comment]
@bertram25 Many thanks, I will find someone to review it before we merge

**Comment by davvd** [comment]
@carlosmiranda this pull request is for you, please review

File: src/main/java/org/takes/rs/RsWithCookie.java [hunk 1]
```diff
@@ -125,6 +145,33 @@ private static String make(final CharSequence name,
...
+        final StringBuilder value = new StringBuilder();
+        try {
+            for (final String header : res.head()) {
+                if (header.contains(RsWithCookie.SET_COOKIE)) {
+                    cookie.append(header);
+                    break;
+                }
+            }
+        } catch (final IOException exception) {
+            return value.toString();
```
Comment (carlosmiranda) [inline src/main/java/org/takes/rs/RsWithCookie.java]: @Bertram25 Why are we swallowing an Exception here and then performing a return?

File: src/test/java/org/takes/rs/RsWithCookieTest.java [hunk 2]
```diff
@@ -38,22 +38,27 @@
 public final class RsWithCookieTest {
 
     /**
-     * RsWithCookie can add cookies.
+     * RsWithCookie can add cookies (even with several decorations).
```
Comment (carlosmiranda) [inline src/test/java/org/takes/rs/RsWithCookieTest.java]: @Bertram25 Let's keep this unit test as it is. Instead, add a second test "addsMultipleCookies" with the behavior we're specifically trying to implement.

File: src/main/java/org/takes/rs/RsWithCookie.java [hunk 3]
```diff
@@ -95,27 +109,33 @@ public RsWithCookie(final Response res, final CharSequence name,
...
      */
-    private static String make(final CharSequence name,
+    private static String make(final String previous, final CharSequence name,
         final CharSequence value, final CharSequence... attrs) {
-        final StringBuilder text = new StringBuilder(
+        final StringBuilder text = new StringBuilder();
+        if (!previous.isEmpty()) {
+            text.append(String.format("%s,", previous));
+        }
+        text.append(
```
Comment (carlosmiranda) [inline src/main/java/org/takes/rs/RsWithCookie.java]: @Bertram25 We can pass `String.format` statement below on this line without reaching the line length limit.

File: src/main/java/org/takes/rs/RsWithCookie.java [hunk 4]
```diff
@@ -125,6 +145,33 @@ private static String make(final CharSequence name,
     }
 
     /**
+     * Retrieve potential previous cookie string.
+     * @param res Current Response
+     * @return Current cookie value or empty string.
+     */
+    @SuppressWarnings("PMD.OnlyOneReturn")
+    private static String getPreviousValue(final Response res) {
```
Comment (carlosmiranda) [inline src/main/java/org/takes/rs/RsWithCookie.java]: @Bertram25 Let's simply call it `previousValue`.

**Comment by carlosmiranda** [comment]
@Bertram25 Some comments above.

**Comment by Bertram25** [comment]
@carlosmiranda 

Thanks for the review. :)

> Let's simply call it previousValue.
We can pass String.format statement below on this line without reaching the line length limit.
Let's keep this unit test as it is. Instead, add a second test "addsMultipleCookies" with the behavior we're specifically trying to implement.

Those points should now be fixed:

> Why are we swallowing an Exception here and then performing a return?

As for this point, I've been thinking about it, and I think you're right it is not the best thing to do.
Here is the situation:
res.head() can throw an IOException. And if I make previousValue() throw an Exception as well, then I'll end up being forced to make the RsWithCookie constuctor capable of throwing an exception.
As I didn't see that being done on any constructors in takes, I think this makes it pretty obvious this isn't the right thing to do.
Thus, if I want to make it use the exception correctly, I guess I'd have to create a new Response tye within the RsWithCookie class overriding the head() function and replacing the cookie header, with the cost that I will no more be able to use the Without/WithHeader decorations in that constructor.

If that's what desired, well, I'm back to square one but I can do it, the feature can also be seen in a more general point of view and maybe it is time to create a RsReplaceHeader class?

Just tell me. :)

Best regards,

**Comment by carlosmiranda** [comment]
@Bertram25 your suggested strategy in overriding `Response` methods is similar to how other classes (e.g. `RsWithHeaders`) does it. Let's do that.

**Comment by Bertram25** [comment]
@carlosmiranda Thanks a lot for the hints given!
The new commit should fix the last remaining issue.

**Comment by Bertram25** [comment]
@davvd Please give me one more day to give time to the review. Thanks!

File: src/main/java/org/takes/rs/RsWithCookie.java [hunk 5]
```diff
@@ -94,30 +114,70 @@ public RsWithCookie(final CharSequence name, final CharSequence value,
...
-                RsWithCookie.make(
-                    RsWithCookie.checkName(name),
-                    RsWithCookie.checkValue(value),
-                    attrs
-            )
-        )
+            new Response() {
+                @Override
+                public Iterable<String> head() throws IOException {
+                    return RsWithCookie.extend(
```
Comment (carlosmiranda) [inline src/main/java/org/takes/rs/RsWithCookie.java]: @Bertram25 I think we can move contents of `extend` method in here now.

**Comment by carlosmiranda** [comment]
@Bertram25 Just one more comment. :)

**Comment by Bertram25** [comment]
@carlosmiranda It should be fixed now.
I had to disable a PMD rule complaining about the lack of a super() call, even though the call is actually done.

**Comment by carlosmiranda** [comment]
@Bertram25 Strange error, but thanks, all looking good now.

**Comment by carlosmiranda** [comment]
@rultor Let's merge please

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by davvd** [comment]
@elenavolokhova review this ticket please, for compliance with our [quality rules](http://www.teamed.io/qa.html)

**Comment by davvd** [comment]
@rultor deploy pls

**Comment by elenavolokhova** [comment]
@carlosmiranda 
 According to our [quality rules](http://www.teamed.io/qa.html):
 > Pull request description explains the solution proposed and contains a link to the original ticket it is related to.
 
 There is poor description provided and it is not linked to original ticket. Pull request must briefly explain what was done in the branch, so that we can use this information later, for traceability. 
As far as you are the performer of this ticket you are responsible for its quality. Please fix this issue and make sure that everyone follows this rule in your tickets in the future.

**Comment by carlosmiranda** [comment]
@Bertram25 please elaborate the PR description and link to the original ticket.

**Comment by Bertram25** [comment]
@carlosmiranda My apologies for not doing this earlier. The mistake is duly noted. Is it better now?

**Comment by elenavolokhova** [comment]
@Bertram25 @carlosmiranda Thank you!

**Comment by elenavolokhova** [comment]
@davvd Quality is acceptable here.

**Comment by davvd** [comment]
> @davvd Quality is acceptable here.

@elenavolokhova thanks for the QA review, we'll work better next time

**Comment by davvd** [comment]
@carlosmiranda just added 10 mins to @elenavolokhova (for QA), payment ID is `58673640`

**21 mins** sent to your balance (ID `58673643`), many thanks! It took 176 hours and 12 mins.

review comments (c=6) added as a bonus

+21 added to your rating, at the moment it is: [+3986](http://www.netbout.com/b/35023?open=rating)

**Pull request merged**